Add policy management services

jjbuchan opened the pull request:
# What

Adds the first implementation of the Policy Management service.

# How

This contains of two main components, `PolicyManagement` and `TenantManagement`.

`TenantManagement` is pretty simple and right now is only really used to store the `AccountType` of a particular tenant, which can be used to determine what Monitor Policies relate to it.

`PolicyManagement` is where all the Monitor Policies are stored.  It creates `PolicyEvent`s whenever a policy is created or deleted (updates are not allowed).

See the design doc for more info.  In future it will reside in https://github.com/racker/salus-docs

## How to test

Run tests

# Why

This will be used to manage default policies and default metadata.

For example, there could be a Global Monitor Policy that causes a Ping monitor to be created for every resource in the system.  That monitor could also be configured to use default metadata values such as `${rackspace.monitoring_zones}`.  Those defaults would reside in Policy Management, and if they were ever updated all bound monitors using it would be updated.

# TODO

Next step is updating Monitor Management to handle the policy events.

Also need to update insomnia.

jjbuchan commented on src/main/java/com/rackspace/salus/policy/manage/services/PolicyManagement.java:
```diff
@@ -0,0 +1,206 @@
+                    Collectors.maxBy(Comparator.comparingInt(policy -> policy.getScope().getPriority())))
+            )
+            // Now we have a map of policy name -> (optional) monitor policy
+            .values().stream()
```
This final part where we stream through the values can probably be optimized so we do it all in the step before.

For the amount of policies that will exist it shouldn't be a big deal being left as is, but if anyone has any ideas in how to optimize it please let me know.  I didn't want to spend much longer on it and the few attempts I made at expanding on the collector grouping/filtering didn't work.

GeorgeJahad replied:
it seems fine to me as it is

itzg commented on src/main/java/com/rackspace/salus/policy/manage/services/PolicyManagement.java:
```diff
@@ -0,0 +1,206 @@
+   * @return True if the monitor exists, otherwise false.
+   */
+  private boolean isValidMonitorId(String monitorId) {
+    return monitorApi.getPolicyMonitorById(monitorId) != null;
```
It would be nice if we had an "exists" API call available here. This won't be frequent enough to be a performance problem, but it's a shame the whole monitor DTO has to come back just to check for null-ness.

jjbuchan replied:
I can change that later.  As mentioned in the other PR, I had considered it already.

jjbuchan replied:
Created a jira for this.

itzg commented on src/main/java/com/rackspace/salus/policy/manage/services/PolicyManagement.java:
```diff
@@ -0,0 +1,206 @@
+   * Sends monitor policy events for all potentially relevant tenants.
+   * @param policy The MonitorPolicy to distribute out to all tenants.
+   */
+  private void sendMonitorPolicyEvents(MonitorPolicy policy) {
```
Should this be more selective and only send events for all tenants only if it's a globally scoped policy? Likewise if it's an account type or tenant scoped policy could selectively send to only applicable tenants.

jjbuchan replied:
That's a good point.

I'll create a jira and optimize that in a second PR.  That will also make use of my idea of being able to get a list of tenants by AccountType.

jjbuchan replied:
Created an issue and added it to this sprint.

itzg commented on src/main/java/com/rackspace/salus/policy/manage/services/TenantManagement.java:
```diff
@@ -0,0 +1,99 @@
+  public String getAccountTypeByTenant(String tenantId) {
+    Optional<TenantMetadata> metadata = getMetadata(tenantId);
+    if (metadata.isEmpty()) {
+      return null;
```
or should it throw a not found?

jjbuchan replied:
Maybe?  Some of this logic I follows what we do in other places, like in API clients where we return `null` if something is not found.

If I change this I'd like to make it a little more consistent everywhere.

jjbuchan replied:
Using NotFound would make it easier to bundle this into the API if we wanted.  Right now it's only used locally so `null` works out.

itzg commented on src/main/java/com/rackspace/salus/policy/manage/web/controller/PolicyApiController.java:
```diff
@@ -0,0 +1,94 @@
+        () -> new NotFoundException(String.format("No policy found with id %s", uuid))).toDTO();
+  }
+
+  @GetMapping("/admin/policy/monitors/effective/{tenantId}")
```
I wonder if "applicable" is a better word than "effective" for this and the management method called. "Effective" makes me think of rendered monitor configs.

jjbuchan replied:
You maybe noticed I've used both those terms in this PR.

I used "applicable" for any policy that relates to an account even if it is not being used due to being overridden by another.  Based on Google telling me "applicable" means "relevant or appropriate" I think that may work.

Google's definition of "effective", "successful in producing a desired or intended result", doesn't really help clear up either of the uses.  To me "effective' is the one's that in the end will have an "effect" on the customer.  Something will actually happen due to it, it will not be ignored.

I'm open to changing things though.  This may be another case for @GeorgeJahad to have the deciding vote (or throw in his own terms).

itzg replied:
I don't feel strongly about either word, so yeah, leave it as is until @GeorgeJahad has a good suggestion.

GeorgeJahad replied:
I don't feel too strongly but applicable is clearer to me.

itzg commented on src/main/java/com/rackspace/salus/policy/manage/web/controller/PolicyApiController.java:
```diff
@@ -0,0 +1,94 @@
+  @PostMapping("/admin/policy/monitors")
+  @ResponseStatus(HttpStatus.CREATED)
+  @ApiOperation(value = "Creates new Monitor for Tenant")
+  @ApiResponses(value = { @ApiResponse(code = 201, message = "Successfully Created Monitor Policy")})
```
Just realized you could probably use `HttpStatus.CREATED.value()` to avoid needing to remember the literal status code of 201, etc.

jjbuchan replied:
Went to change things, but it doesn't like it

<img width="283" alt="image" src="https://user-images.githubusercontent.com/4358210/62229728-9397e500-b37d-11e9-99c2-32509cf892c3.png">

itzg replied:
Bummer, thanks for trying

itzg commented on src/main/java/com/rackspace/salus/policy/manage/web/controller/RestExceptionHandler.java:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  @ExceptionHandler({IllegalArgumentException.class})
+  public ResponseEntity<?> handleBadRequest(
```
I think we repeat these three handlers each time. One of these days we should create a `DefaultRestExceptionHandler` so that these concrete ones can typically be empty classes to declare the `@ControllerAdvice`.

itzg commented on src/main/java/com/rackspace/salus/policy/manage/web/controller/TenantApiController.java:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  @GetMapping("/public/account/{tenantId}")
+  @ApiOperation(value = "Retrieves miscellaneous information stored for a particular tenant")
```
"miscellaneous" seems vague, but then I guess that's the nature of metadata.

jjbuchan replied:
Yep :)

itzg commented on src/test/java/com/rackspace/salus/policy/manage/services/PolicyManagementTest.java:
```diff
@@ -0,0 +1,422 @@
+        .setName(defaultMonitorPolicy.getName())
+        .setMonitorId(RandomStringUtils.randomAlphabetic(10));
+
+    assertThatThrownBy(() -> policyManagement.createMonitorPolicy(policyCreate))
```
Cool, AssertJ does have that

jjbuchan replied:
Yep, it had a few different versions.  

Down below I could have used the shortcut `assertThatIllegalArgumentException()`, but sticking with consistent methods seemed better.

itzg commented on src/test/java/com/rackspace/salus/policy/manage/services/TenantManagementTest.java:
```diff
@@ -0,0 +1,126 @@
+  @Test
+  public void testGetMetadata() {
+    TenantMetadata original = tenantMetadataRepository.save(podamFactory.manufacturePojo(TenantMetadata.class));
+    tenantManagement.getMetadata(original.getTenantId());
```
Is this an extra call?

jjbuchan replied:
It is.  Removed.

jjbuchan commented on pom.xml:
```diff
@@ -0,0 +1,283 @@
+    </dependency>
+    <dependency>
+      <groupId>com.rackspace.salus</groupId>
+      <artifactId>salus-telemetry-protocol</artifactId>
```
I don't think it does.  As always I copied the initial version of this from somewhere else then tried to clean it up.  This got missed.

jjbuchan commented on src/main/java/com/rackspace/salus/policy/manage/entities/MonitorPolicy.java:
```diff
@@ -0,0 +1,52 @@
+    return new MonitorPolicyDTO()
+        .setMonitorId(monitorId)
+        .setName(name)
+        .setSubscope(subscope)
```
That could be a good idea.  I'll look into it.

jjbuchan replied:
Unless I'm missing something, I don't think we can do that.

itzg replied:
Even so, it's probably an uglier refactoring than the problem it could solve. So, I'm good as is anyway.

jjbuchan commented on src/main/java/com/rackspace/salus/policy/manage/services/PolicyManagement.java:
```diff
@@ -0,0 +1,206 @@
+        // Create a stream from all monitor policies
+        StreamSupport.stream(monitorPolicyRepository.findAll().spliterator(), false)
+            // Filter the stream for only those policies relevant to this tenant
+            .filter(policy -> isPolicyApplicable(policy, tenantId))
```
Definitely agree on all points.  I really don't think we'll have enough policies for performance to be a big problem.

jjbuchan commented on src/main/java/com/rackspace/salus/policy/manage/services/PolicyManagement.java:
```diff
@@ -0,0 +1,206 @@
+            // Filter out any of the optional objects that do not contain anything
+            .filter(Optional::isPresent).map(Optional::get)
+            // Finally convert to a list of the relevant monitor policies
+            .collect(Collectors.toList()));
```
Nope, I'll remove that.  It's leftover from trying some of IntelliJ's recommendations for simplifying the query.

jjbuchan replied:
Done

jjbuchan commented on src/main/java/com/rackspace/salus/policy/manage/web/model/MonitorPolicyDTO.java:
```diff
@@ -0,0 +1,25 @@
+
+import lombok.Data;
+
+@Data
```
So it does.  I never noticed that.

jjbuchan commented on pom.xml:
```diff
@@ -0,0 +1,283 @@
+    </dependency>
+    <dependency>
+      <groupId>com.rackspace.salus</groupId>
+      <artifactId>salus-telemetry-monitor-management</artifactId>
```
You're right.  I thought we might have something else like that, but I can't think of anything now.

We're lucky they use each other in entirely different flows, so we don't have a real cyclical problem.

I think we'd have to switch to the method of including more info on `PolicyEvent` if we ever had to get away from this.  It'd likely have to have a `MonitorDTO` object on there.  That would also mean moving those objects out of mon-mgmt and over to salus-models.

jjbuchan commented on src/main/java/com/rackspace/salus/policy/manage/entities/TenantMetadata.java:
```diff
@@ -0,0 +1,77 @@
+
+  @NotBlank
+  @Column(name = "tenant_id", unique = true)
+  String tenantId;
```
Done

jjbuchan commented:
@itzg I think I've addressed all your comments.  The ones I haven't done anything for seem like ones that should be changed across all repos if we change them vs. just in here, so i'll leave them alone for now.

itzg commented on src/main/java/com/rackspace/salus/policy/manage/entities/TenantMetadata.java:
```diff
@@ -0,0 +1,80 @@
+
+@Entity
+@Table(name = "tenant_metadata", indexes = {
+    @Index(name = "by_tenant", columnList = "tenant_id")
```
I'm never quite sure if `@Index` needs the DB column name `tenant_id` or the JPA field name `tenantId`, so heads up this may or may not work with `tenant_id`.

jjbuchan replied:
From what I've seen I think it's the column name.  I assume I can see that when I run everything in dev.  I'll try remember to have a look later.

itzg replied:
Yeah, it'll fail fast if it's not the right one of the two choices.

itzg commented on pom.xml:
```diff
@@ -0,0 +1,278 @@
+    <relativePath>../salus-app-base</relativePath>
+  </parent>
+
+  <artifactId>salus-telemetry-policy-management</artifactId>
```
All of the telemetry parts of Salus (things "to the left" of UMB) have telemetry in their artifact ID except for salus-common. Event engine artifacts are not part of telemetry and don't have "-telemetry-", such as

https://github.com/racker/salus-event-engine-common/blob/master/pom.xml#L24

Policy management ultimately benefits both telemetry and event-engine parts of Salus, so I could go either way with or without the "-telemetry-" qualifier in the ID.

jjbuchan replied:
It was updating the docker config that made me notice this.  I'll leave it how things are now.

itzg replied:
Yeah, I see what you mean. Using my logic above, public-api and admin-api probably shouldn't have "-telemetry-" either :)
